addons/event_crm/models/event_registration.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+
+
+class EventRegistration(models.Model):
+    _inherit = 'event.registration'
+
+    lead_ids = fields.Many2many('crm.lead', string='Leads', readonly=True, help="Leads generated from the registration.", copy=False)
+    leads_count = fields.Integer('# Leads', compute='_compute_leads_count', help="Counter for the leads linked to this registration")
+
+    @api.depends('lead_ids')
+    def _compute_leads_count(self):
+        for record in self:
+            record.leads_count = len(record.lead_ids)
+
+    @api.model
+    def create(self, vals):
+        registration = super(EventRegistration, self).create(vals)
+        self.env['event.crm.rule'].search([])._create_leads(registration)
+        return registration
+
+    def write(self, vals):
+        registration = super(EventRegistration, self).write(vals)
+        for lead in self.lead_ids:
+            vals = self._get_registration_lead_values(lead.registration_rule_id.lead_creation_basis, lead.registration_ids)
+            lead.write(vals)
+        return registration
+
+    def _get_registration_lead_values(self, rule_type, registration_ids=False):
+        registration_lead_values = {
+            'event_id': self.event_id.id,
+            'referred': self.event_id.name,
+        }
+        if rule_type == 'attendee':
+            registration = registration_ids if registration_ids else self
+            registration = self.search([('id', '=', registration.id)])
+            description = _("Participant:\n")
+            description += registration._get_lead_description()
+            registration_lead_values.update({
+                'name': "%s - %s" % (self.event_id.name, registration.name),
+                'contact_name': registration.name,
+                'email_from': registration.email,
+                'phone': registration.phone,
+                'mobile': registration.mobile,
+                'registration_ids': registration.ids,
+                'description': description,
+            })
+            if registration.partner_id and registration.partner_id != self.env.ref('base.public_partner'):
+                if not registration.main_registration_id:
+                    registration_lead_values.update({
+                        'name': "%s - %s" % (self.event_id.name, registration.partner_id.name),
+                        'partner_id': registration.partner_id.id,
+                        'email_from': registration.partner_id.email,
```

Comment:
You should first set the partner (and no other info as computed field will populate other information), and if not set other fields. That way you will have less code and rely more on existing computed fields :) .